projects/lexical_shared/lib/lexical/document/store.ex

```diff
@@ -100,123 +151,101 @@ defmodule Lexical.Document.Store do
-      ets_delete(uri, :temp)
-      %__MODULE__{store | temporary_open_refs: new_refs}
-    end
-
-    defp maybe_cancel_old_ref(%__MODULE__{} = store, uri) do
-      {_, new_refs} =
-        Map.get_and_update(store.temporary_open_refs, uri, fn
-          nil ->
-            :pop
-
-          old_ref when is_reference(old_ref) ->
-            Process.cancel_timer(old_ref)
-            :pop
-        end)
-
-      new_refs
+      {_, store} = pop_open_doc(store, uri)
+      maybe_cancel_ref(store, uri)
     end
 
-    defp schedule_unload(uri, timeout) do
-      Process.send_after(self(), {:unload, uri}, timeout)
+    defp put_open_doc(%__MODULE__{} = store, %Document{} = document, derived \\ %{}) do
+      put_in(store.open[document.uri], open_doc(document: document, derived: derived))
     end
 
-    defp normalize_error(:error), do: {:error, :not_open}
-    defp normalize_error(e), do: e
-
-    @read_types [:sources, :temp, :any]
-    @write_types [:sources, :temp]
-    defp ets_fetch(key, type) when type in @read_types do
-      case :ets.match(@table_name, {key, type_selector(type), :"$1"}) do
-        [[value]] -> {:ok, value}
-        _ -> :error
+    defp pop_open_doc(%__MODULE__{} = store, uri) do
+      case Map.pop(store.open, uri) do
+        {open_doc() = doc, open} -> {doc, %__MODULE__{store | open: open}}
+        {nil, _} -> {nil, store}
       end
     end
 
-    defp ets_put(key, type, value) when type in @write_types do
-      :ets.insert(@table_name, {key, type, value})
-      :ok
+    defp put_ref(%__MODULE__{} = store, uri, ref) do
+      put_in(store.temporary_open_refs[uri], ref)
     end
 
-    defp ets_has_key?(key, type) when type in @read_types do
-      match_spec = {key, type_selector(type), :"$1"}
-
-      case :ets.match(@table_name, match_spec) do
-        [] -> false
-        _ -> true
-      end
-    end
+    defp maybe_cancel_ref(%__MODULE__{} = store, uri) do
+      case pop_in(store.temporary_open_refs[uri]) do
+        {ref, store} when is_reference(ref) ->
+          Process.cancel_timer(ref)
+          store
 
-    defp ets_pop(key, type) when type in @write_types do
-      with {:ok, value} <- ets_fetch(key, type),
-           :ok <- ets_delete(key, type) do
-        value
-      else
         _ ->
-          nil
+          store
       end
     end
 
-    defp ets_delete(key, type) when type in @write_types do
-      match_spec = {key, type, :_}
-      :ets.match_delete(@table_name, match_spec)
-      :ok
+    defp schedule_unload(uri, timeout) do
+      Process.send_after(self(), {:unload, uri}, timeout)
     end
 
-    defp type_selector(:any), do: :_
-    defp type_selector(type), do: type
-  end
-
-  alias Lexical.Document
-  alias Lexical.ProcessCache
+    defp normalize_error(:error), do: {:error, :not_open}
+    defp normalize_error(e), do: e
 
-  @type t :: %State{}
+    defp derive(%__MODULE__{} = store, key, document) do
+      case store.derivation_funs do
+        %{^key => fun} ->
```

Comment:
maybe: 
```suggestion
        %{^key => fun} when is_function(fun, 1) ->
```

Ideally, this check should be done in the constructor